usr/share/rear/conf/default.conf

```diff
@@ -3289,39 +3289,71 @@ SIMPLIFY_TEAMING=no
 
 ####
 # Serial console support for the ReaR rescue/recovery system:
-# By default serial consoles get enabled if serial devices are found and
-# then matching kernel command line parameters like 'console=ttyS0,9600 console=ttyS1,9600'
-# are set when booting the rescue/recovery system (see KERNEL_CMDLINE above).
 # IA64 platforms require it and people need it when no VGA console is available.
-# Specify 'yes' or 'no' or leave it empty to use ReaR's automatism:
+# By default there is serial console support in the ReaR recovery system
+# when 'getty' or 'agetty' and 'stty' can be found,
+# otherwise USE_SERIAL_CONSOLE is automatically set to 'no'.
+# When 'getty' or 'agetty' and 'stty' can be found
+# and there is a 'console=...' option in /proc/cmdline,
+# then USE_SERIAL_CONSOLE is automatically set to 'yes'.
+# With USE_SERIAL_CONSOLE="no" no serial console gets set up,
+# neither for the recovery system kernel nor for the recovery system bootloader.
+# With USE_SERIAL_CONSOLE="yes" plus appropriate SERIAL_CONSOLE_DEVICE... settings
+# serial consoles can be specified for the recovery system kernel and bootloader
+# for example when there is no 'console=...' option in /proc/cmdline
+# or when serial consoles for the recovery system kernel and bootloader
+# should differ from what 'console=...' options in /proc/cmdline tell.
+# By default (when empty) the below further described automatism is used:
 USE_SERIAL_CONSOLE=
-# If you do not want to use all serial devices found as console you can specify the ones to use
-# e.g. SERIAL_CONSOLE_DEVICES="/dev/ttyS0 /dev/ttyS1" provided USE_SERIAL_CONSOLE is turned on.
-# By default (when empty) all serial devices found are used:
+#
+# Which serial devices should be used for serial consoles
+# provided USE_SERIAL_CONSOLE is not set to 'no':
+# E.g. SERIAL_CONSOLE_DEVICES="/dev/ttyS0 /dev/ttyS1"
+# By default (when empty) the devices of the 'console=...' options in /proc/cmdline
+# that exist as /dev/ttyS* or /dev/hvsi* character device nodes are used
+# (which excludes /dev/tty0 when there is 'console=tty0' in /proc/cmdline):
 SERIAL_CONSOLE_DEVICES=
-# Serial consoles for the kernel of the recovery system:
+#
+# Serial consoles for the kernel of the recovery system
+# provided USE_SERIAL_CONSOLE is not set to 'no':
```

Comment:
@pcahyna

does this need a similar correction as the bootloader parameter descriptions below?

@jsmeix

No.

See the code in
prep/GNU/Linux/200_include_serial_console.sh
here in this pull request.

@pcahyna

isn't it confusing though? At one spot, we check for the value not being false at another spot we check for it being true (which is not equivalent due to the ternary nature of the parameter).
If this was the former behavior, I understand the desire to not introduce a behavior change, but perhaps this is a change to consider in (backward-incompatible) ReaR 3.0 ?

@jsmeix

It is what the code does.
The code does what (at least for me) seems most appropriate
for the usual cases which are
either no 'console=...' option in /proc/cmdline
or one or more 'console=...' options in /proc/cmdline

Therefore the default behaviour depends on that
so the code in prep/GNU/Linux/200_include_serial_console.sh
checks and does the kernel stuff first, and
if a 'console=...' option is set for the recovery system kernel
it sets USE_SERIAL_CONSOLE="yes" which later triggers
serial console setup for the recovery system bootloader.

@jsmeix

For users who want to set up their own stuff
all should be simple and straightforward
with USE_SERIAL_CONSOLE="yes" plus
appropriate SERIAL_CONSOLE_DEVICE... settings

@jsmeix

That USE_SERIAL_CONSOLE="yes" is needed for
serial console setup for the recovery system bootloader
is existing code that I do not want to change with this pull request
because this pull request is only about to
"Copy the console= kernel arguments from the original system"

Reply:
After sleeping on it I even think now
that the logic is perfectly right:

When USE_SERIAL_CONSOLE is empty
then the code in prep/GNU/Linux/200_include_serial_console.sh
(the first ...console... script that is run for "rear mkrescue")
decides if USE_SERIAL_CONSOLE is kept empty
or (provided there is sufficient reason)
it sets it to 'no' or 'yes'.

'no' when serial consoles cannot work in the recovery system
and
'yes' when a serial console will be set up for the
recovery system kernel.

Only when a serial console will be actually set up
for the recovery system kernel, then it makes sense
to also set up a serial console will be set up
for the recovery system bootloader.

So the serial console set up for the recovery system
bootloader is only done if USE_SERIAL_CONSOLE="yes"
because an empty USE_SERIAL_CONSOLE must not result
some serial console set up for the recovery system bootloader
(without serial console set up for the recovery system kernel).

Reply:
I think I will add my explanation in
https://github.com/rear/rear/pull/2961#discussion_r1358150770
as comment to prep/GNU/Linux/200_include_serial_console.sh
so other developers understand why the code is as it is, cf.
https://github.com/rear/rear/wiki/Coding-Style#code-must-be-easy-to-understand-answer-the-why

But I would not like to have such explanations
about the implementation in default.conf

Reply:
Added via
https://github.com/rear/rear/pull/2961/commits/9670c4dfe84c3958e0fd5af32fa825268c9836d7

Reply:
^ originally from https://github.com/rear/rear/commit/478355980e80eecab90fe586f2ce7a49738c9839#r129706534

Reply:
@jsmeix

> That USE_SERIAL_CONSOLE="yes" is needed for
serial console setup for the recovery system bootloader
is existing code that I do not want to change with this pull request
because this pull request is only about to
"Copy the console= kernel arguments from the original system"

I fully understand the motivation of not wanting to change behavior except when it is needed to fix the bug that motivated this change. For the future (ReaR 3.0) though, I would like to think again about the semantics and possibly change it. In particular, I don't like that there are conditions stated as
> provided USE_SERIAL_CONSOLE is not set to 'no':

while other conditions are stated as:
> provided USE_SERIAL_CONSOLE is 'yes':

which makes harder to understand the impact of settings when reading default.conf (the same empty value has a different consequence in one place than in another place). Would it make sense to change it to "provided USE_SERIAL_CONSOLE is 'yes':" everywhere (and also in the code) in a future version, for consistency and simplicity?

> After sleeping on it I even think now
that the logic is perfectly right:

to me that seems to explain why the condition for the bootloader is "provided USE_SERIAL_CONSOLE is 'yes':" but not why the condition for the kernel part is "provided USE_SERIAL_CONSOLE is not set to 'no':".

Reply:
The explanation why the condition for the kernel part is
"provided USE_SERIAL_CONSOLE is not set to 'no':"
is here
https://github.com/rear/rear/pull/2961/files#r1351973906
that reads (I copy it here to be safe against it may get lost)
```
When USE_SERIAL_CONSOLE=no
kernel command line console= parameters on the original system
will not be copied as kernel command line parameters
for the kernel of the ReaR recovery system.
This is needed when the user specifies USE_SERIAL_CONSOLE=no
because then ReaR has to obey.
```
For users who do not want that ReaR does serial console setup
plain USE_SERIAL_CONSOLE="no" should do it.

Reply:
In
https://github.com/rear/rear/pull/2961/commits/6d410a2fda92d763bc2f27736ae37bd5b144460a
I simplified the wording from
"provided USE_SERIAL_CONSOLE is not set to 'no'"
to
"provided USE_SERIAL_CONSOLE is not 'no'"
and I made the automated USE_SERIAL_CONSOLE='yes' more clear as
"provided USE_SERIAL_CONSOLE is 'yes' (automatically with 'console=...' in /proc/cmdline)"